DOC-11961 Provide comprehensive documentation for hotspots

florence-crl opened the pull request:
Fixes DOC-11961

Added understand-hotspots.md and corresponding images.
In sidebar-data/troubleshooting.json, added link to understand-hotspots.md.

Rendered preview:
- [Understand Hotspots](https://deploy-preview-19282--cockroachdb-docs.netlify.app/docs/v25.1/understand-hotspots.html)

florence-crl commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,308 @@
+
+For example, consider the following schema:
+
+~~~ sql
+CREATE TABLE products (
+   id SERIAL PRIMARY KEY USING HASH
+);
+~~~
```
@angles-n-daemons , would an example using a `SEQUENCE` rather than `SERIAL` be clearer? Something like:

~~~ sql
CREATE TABLE products (
    id INT PRIMARY KEY USING HASH,
    name STRING
);
CREATE SEQUENCE products_id_seq;
--Insert 2 million rows
INSERT INTO products (id, name) SELECT nextval('products_id_seq'), (1000000000000000 + generate_series(1,2000000))::STRING;
~~~

angles-n-daemons replied:
It would, actually this is a good note. It turns out that `SERIAL` doesn't use sequences in CRDB as it does in postgres. Can you make this change?

florence-crl replied:
replaced `SERIAL` example with `SEQUENCE` example.

angles-n-daemons commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,308 @@
+
+**Synonyms:** hot key, hot row, point hotspot
+
+A _row hotspot_ is a hotspot where an individual point in the keyspace becomes throughput limiting, often due to high activity on a single row. The word "point" illustrates the fact that the location that is hot is indivisible, and therefore cannot be [split]({% link {{ page.version.version }}/load-based-splitting.md %}). A classic example is a social media application with a high volume of activ ...
```
Since there isn't reference to point hotspots before this point, might be worth scrapping the reference:

`Since rows are inherently indivisible, a hotspot on a row cannot be split...`

angles-n-daemons reviewed (commented):
This is great, left a couple comments - but appreciate all the extra work put into this (and the links are amazing).

florence-crl reviewed (commented):
tftr! @kevin-v-ngo and @angles-n-daemons

angles-n-daemons commented:
Florence one more thing to note, the hot key visual may be misleading. While write hotspots can go up to 20k writes per second for an index hotspot, hot keys are generally limited to < 1k writes per second. The image I created seems to indicate that they can go much higher, which I doubt is the case in most deployments. Not sure if this is important for documentation, but it might be worth updating the image. I've updated it in the "Talking about hotspots" document

florence-crl commented:
> Florence one more thing to note, the hot key visual may be misleading. While write hotspots can go up to 20k writes per second for an index hotspot, hot keys are generally limited to < 1k writes per second. The image I created seems to indicate that they can go much higher, which I doubt is the case in most deployments. Not sure if this is important for documentation, but it might be worth updating the image. I've updated it in the "Talking about hotspots" document

I updated the 2 images.

kevin-v-ngo commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,312 @@
+
+_Temporal hotspots_ refer to increased database usage during particular windows of time. These take a variety of shapes, from event and holiday usage (such as Black Friday or the Superbowl), to synchronized job runs.
+
+### Internal Hotspot
```
I'm not sure we should call this out since it's uncommon - what the guidance to troubleshoot this scenario?

florence-crl replied:
Removed section:

```
### Internal Hotspot

**Synonyms:** hot job, task hotspot

An _internal hotspot_ refers to a type of hotspot that arises from the internal operations of CockroachDB. These hotspots are significant because, although they may be caused by operator error, they often appear in ranges within the keyspace that operators are not familiar with. Examples of internal hotspots include activities such as metrics writing/scraping and jobs such as SQL statistics generation. These internal operations can create bottlenecks and performance issues within the cluster.
```

kevin-v-ngo commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,312 @@
+
+### Load Balancing Hotspot
+
+A _load balancing hotspot_ is a hotspot caused by a load balancer misconfiguration. Although this issue is often considered an afterthought due to its low internal tracking value, it is important to note that all possible consumer groups can create uneven load on the cluster. For example, a [changefeed]({% link {{ page.version.version }}/change-data-capture-overview.md %}) subscription can create ...
```
What low internal tracking value?

florence-crl replied:
@angles-n-daemons, what did you mean by the clause below?

> Although this issue is often considered an afterthought due to its low internal tracking value

angles-n-daemons replied:
I'm not necessarily sure - I would say it's considered an afterthought because key related hotspots are generally more common.

florence-crl replied:
updated

kevin-v-ngo reviewed (changes requested):
Thanks for putting this together! Few comments.

florence-crl commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,312 @@
+toc: true
+---
+
+Hotspots in distributed SQL are a common topic. The term _hotspot_ is useful to describe bottlenecks in scaling a database cluster within workloads. This page defines terms and patterns to improve understanding for troubleshooting hotspots.
```
Good catch! I was meaning to correct that and forgot. Updated.

florence-crl commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,312 @@
+
+All hotspot types described in this page will create hot nodes, as long as the cluster is not already operating at maximum capacity.
+
+The following image is a graph of [CPU Percent]({% link {{ page.version.version }}/ui-hardware-dashboard.md %}#cpu-percent) utilization by node. Most of the nodes hover around 25%, while one hot node is around 95%. The fact that the hot node changes indicates that the hotspot is moving as the [ranges]({% link {{ page.version.version }}/architecture/overview.md %}#range) containing writes fill up  ...
```
added links to subsequent sections.

florence-crl commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,312 @@
+
+If a node is hot, it is often due to a single hot range. The system may split the hot range to redistribute the load or the range may stay hot until it fills up and splits. In the second case, the split is likely the continuation of the hotspot (as shown in the previous image). If the system is unable to identify a good splitting point for a hot range, the hot range becomes a bottleneck.
+
+Understanding which range is hot and having knowledge of the [keyspace]({% link {{ page.version.version }}/architecture/overview.md %}#range) allows you to better approximate the nature of the hotspot compared to relying solely on node-level statistics.
```
On the [DB Console Hot Ranges page](https://www.cockroachlabs.com/docs/v25.1/ui-hot-ranges-page) , it says:
> The Hot Ranges page of the DB Console provides details about ranges receiving a high number of reads or writes. These are known as hot ranges.

A cluster will always have ranges that have a relatively higher number of reads or writes. There are always ranges displayed on the DB Console Hot Ranges page.

 I will add that "Hot ranges are an issue when they cause a hot node."

florence-crl commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,312 @@
+
+**Synonyms**: hot index, golden keyspace hotspot, monotonically increasing index, a running tail, a moving tail
+
+An _index hotspot_ is a hotspot on an [index]({% link {{ page.version.version }}/indexes.md %}) where the key for writes is continually increasing. This is common with indexing by an increasing column (for example, with data type of [`SERIAL`]({% link {{ page.version.version }}/serial.md %}), [`TIMESTAMP`]({% link {{ page.version.version }}/timestamp.md %}), or [`AUTO_INCREMENT`]({% link {{ page. ...
```
updated

florence-crl commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,312 @@
+
+#### Paradoxical Performance Implications
+
+A note on the index hotspot is that they paradoxically increase the efficiency of CockroachDB from the standpoint of the number of CPU cycles per write. This is primarily due to [write compactions]({% link {{ page.version.version }}/architecture/storage-layer.md %}#compaction). As index hotspots continually move in a single direction, the [SST (Sorted String Table)]({% link {{ page.version.versio ...
```
@angles-n-daemons, Would you be able to answer @kevin-v-ngo questions above?

angles-n-daemons replied:
I think they should just know, that the performance benefits have a cliff. They may choose to use hotspot behavior because of the performance enhancements, and it should be known its a short-sighted choice.

florence-crl reviewed (commented):
@kevin-v-ngo updated according to your comments. tftr
@angles-n-daemons, please provide clarification to the unresolved conversations.

kevin-v-ngo commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,311 @@
+
+_Temporal hotspots_ refer to increased database usage during particular windows of time. These take a variety of shapes, from event and holiday usage (such as Black Friday or the Superbowl), to synchronized job runs.
+
+### Load Balancing Hotspot
```
Can we call out the scenario where there could be a load balancer hot spot? One node receiving more connections than others. Connections should be distributed equally across all the nodes in the cluster.

Load balancer configuration issue and should be revisited

florence-crl replied:
updated

kevin-v-ngo reviewed (approved):
Nice! Approved pending small change. Thanks Florence!

kathancox commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+_Pressure_ describes how close resources are to their limits and can be thought of as the force exerted on the actual resources in the system. Pressure can be directed towards disk I/O, CPU, memory, and even non-hardware resources like [locks]({% link {{ page.version.version }}/architecture/transaction-layer.md %}#reading).
+
+It is especially useful when correlating pressure with activity, for example, the read pressure on Node 1 is coming from the key `/User/10`. Pressure is a useful term to describe resource limits, because it captures the ideas of [contention]({% link {{ page.version.version }}/performance-best-practices-overview.md %}#transaction-contention), utilization, and throughput - all of which can be the s ...
```
```suggestion
It is especially useful when correlating pressure with activity, for example, the read pressure on Node 1 is coming from the key `/User/10`. Pressure is a useful term to describe resource limits, because it includes [contention]({% link {{ page.version.version }}/performance-best-practices-overview.md %}#transaction-contention), utilization, and throughput—all of which can be the source of a hotspot.
```

kathancox commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+**Synonyms:** Index Maximum Present Key
+
+An _index tail_ is the largest existing key of an index, lexigraphically. Note that this is different from the index (such as `/Table/1/1`) or index spans (such as `[/Table/1/1, /Table/1/2)`), since these are not valid row keys. An index tail instead refers to the last real key which appears in the index. For example, the head and tail of the index `/Table/1/1` are: `[/Table/1/1/aardvark, /Table/ ...
```
```suggestion
An _index tail_ is the largest existing key of an index, lexigraphically. Note that this is different from the index (such as `/Table/1/1`) or index spans (such as `[/Table/1/1, /Table/1/2)`), because these are not valid row keys. An index tail instead refers to the last real key that appears in the index. For example, the head and tail of the index `/Table/1/1` are: `[/Table/1/1/aardvark, /Table/1/1/zultan]` 
```

kathancox commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+## Patterns
+
+This section goes into detail about workload patterns which result in hotspots.
```
```suggestion
This section goes into detail about workload patterns that can result in hotspots.
```
Will they definitely result in hotspots?

kathancox commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+**Synonyms**: hot index, golden keyspace hotspot, monotonically increasing index, a running tail, a moving tail
+
+An _index hotspot_ is a hotspot on an [index]({% link {{ page.version.version }}/indexes.md %}) where the key for writes is continually increasing. This is common with indexing by an increasing column. For example, the column may be of data type [`SERIAL`]({% link {{ page.version.version }}/serial.md %}), [`TIMESTAMP`]({% link {{ page.version.version }}/timestamp.md %}), or [`AUTO_INCREMENT`]({%  ...
```
```suggestion
An _index hotspot_ is a hotspot on an [index]({% link {{ page.version.version }}/indexes.md %}) where the key for writes is continually increasing. This is common with indexing by an increasing column. For example, the column may be of data type [`SERIAL`]({% link {{ page.version.version }}/serial.md %}), [`TIMESTAMP`]({% link {{ page.version.version }}/timestamp.md %}), or [`AUTO_INCREMENT`]({% link {{ page.version.version }}/serial.md %}#auto-incrementing-is-not-always-sequential). However, an index hotspot is not always determined by the data type. If sequential data generated by the application is inserted into an index, a hotspot may occur. Index hotspots limit horizontal scaling as the index acts as a bottleneck.
```

kathancox commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+An _index hotspot_ is a hotspot on an [index]({% link {{ page.version.version }}/indexes.md %}) where the key for writes is continually increasing. This is common with indexing by an increasing column. For example, the column may be of data type [`SERIAL`]({% link {{ page.version.version }}/serial.md %}), [`TIMESTAMP`]({% link {{ page.version.version }}/timestamp.md %}), or [`AUTO_INCREMENT`]({%  ...
+
+Consider a table `users` which contains a [primary key]({% link {{ page.version.version }}/primary-key.md %}) `user_id` that is an incrementing integer value. Each new key will be the current maximum key + 1. In this way, all writes appear at the index tail. The following image visualizes writes to the `users` table using an incrementing `INT` primary key. Note how all writes are focused at the t ...
```
```suggestion
Consider a table `users` that contains a [primary key]({% link {{ page.version.version }}/primary-key.md %}) `user_id`, which is an incrementing integer value. Each new key will be the current maximum key + 1. In this way, all writes appear at the index tail. The following image visualizes writes to the `users` table using an incrementing `INT` primary key. Note how all writes are focused at the tail of the index, represented by the red section in Range 4.
```

kathancox commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+<img src="{{ 'images/v25.1/hotspots-figure-3.png' | relative_url }}" alt="incrementing INT primary key" style="border:1px solid #eee;max-width:100%" />
+
+If you are able to control for the performance degradation that occurs in Range 4 in this case, you do not change the fact that the system is now limited by how many writes a single range can execute. This then, in theory, limits CockroachDB to the performance of a single node, which is, by nature, antithetical to the goals of a distributed database.
```
I found the first sentence a little hard to parse, does it mean "even you do x, you won't be able to do y"?
```suggestion
Even if you are able to control for the performance degradation that occurs in Range 4, it isn't possible to change that the system is now limited by how many writes a single range can execute. As a result, CockroachDB could be limited to the performance of a single node.
```

I think the final sentence here could be more direct. I gave a suggestion a shot, but I'm not sure if the meaning is right here.

kathancox commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+Index hotspots themselves are [hot by write](#write-hotspot), [moving hotspots](#moving-hotspot) which will continue to bounce around from range to range as the data grows.
+
+In the ideal operation of a distributed SQL database, inserts into an index should be distributed evenly in the index's span to avoid this type of hotspot. Each range responsible for storing keys receives an even share of the workload. If the ranges containing the data become hot in this mode, scaling the cluster is as easy as adding machines horizontally.
```
What do you think about reworking the "scaling the cluster is as easy as adding machine horizontally."? I'm not sure that that is super easy, perhaps you could mention horizontal scaling and link to a reference?

kathancox commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+<img src="{{ 'images/v25.1/hotspots-figure-2.png' | relative_url }}" alt="UUID primary key" style="border:1px solid #eee;max-width:100%" />
+
+Inserts are not the only way that index hotspots can occur. Consider the same table `users` which now has a [secondary index]({% link {{ page.version.version }}/schema-design-indexes.md %}) on a `TIMESTAMP` column:
```
```suggestion
Inserts are not the only way that index hotspots can occur. Consider the same table `users` that now has a [secondary index]({% link {{ page.version.version }}/schema-design-indexes.md %}) on a `TIMESTAMP` column:
```

kathancox commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+UPDATE profiles SET last_seen_at = CURRENT_TIMESTAMP() WHERE user_uuid='d9efa555-e8fd-4793-9239-f484454980cc';
+~~~
+
+Because [`CURRENT_TIMESTAMP()`]({% link {{ page.version.version }}/functions-and-operators.md %}#date-and-time-functions) is a steadily increasing value, this [`UPDATE`]({% link {{ page.version.version }}/update.md %}) will similarly burden the range at the tail of the index. While hotspots on an index tail tend to be the most common, bottlenecks on the head are not unheard of. For example, index ...
```
Could you rephrase "are not unheard of."? It could be a bit casual and also hard to localize. Maybe "are also possible."?

kathancox commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+Now the writes are distributed into the tails of the shards, rather than the tail of the whole index. This benefits write performance but makes reads more challenging. If you need to read a subset of the data, you will have to scan each shard of the index.
+
+#### Paradoxical Performance Implications
+
+A note on the index hotspot is that they paradoxically increase the efficiency of CockroachDB from the standpoint of the number of CPU cycles per write. This is primarily due to [write compactions]({% link {{ page.version.version }}/architecture/storage-layer.md %}#compaction). As index hotspots continually move in a single direction, the [SST (sorted string table)]({% link {{ page.version.versio ...
```
What do you think about a note for this?

florence-crl replied:
updated

kathancox commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+A _row hotspot_ is a hotspot where an individual point in the keyspace becomes throughput limiting, often due to high activity on a single row. Since rows are inherently indivisible, a hotspot on a row cannot be [split]({% link {{ page.version.version }}/load-based-splitting.md %}). A classic example is a social media application with a high volume of activity on certain users.
+
+Consider a social media application which initially consists of two tables, `users` and `follows`:
```
```suggestion
Consider a social media application that initially consists of two tables, `users` and `follows`:
```

florence-crl replied:
updated

kathancox commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+UPDATE User SET follower_count = follower_count+1 WHERE id=2;
+~~~
+
+This simple design works well until it encounters an unexpected surge in activity. For example, consider user 471, who suddenly goes viral and gains millions of followers within an hour. This sudden increase in followers causes a significant amount of write traffic to the range responsible for this user, which the system may not be able to handle efficiently. The following image visualizes a hot  ...
```
```suggestion
This simple design works well until it encounters an unexpected surge in activity. For example, consider user 471, who suddenly gains millions of followers within an hour. This sudden increase in followers causes a significant amount of write traffic to the range responsible for this user, which the system may not be able to handle efficiently. The following image visualizes a hot row in the keyspace. Note how writes are focused on a single point, which cannot be split.
```

florence-crl replied:
updated

kathancox commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+<img src="{{ 'images/v25.1/hotspots-figure-6.png' | relative_url }}" alt="Single row hotspot example" style="border:1px solid #eee;max-width:100%" />
+
+Without changing the default behavior of the system, the load will not be distributed because it needs to be served by a single range. This behavior is not just temporary; certain users, such as celebrities and influencers, may consistently experience a high volume of activity compared to the average user. This can result in a system with multiple hotspots, each of which can potentially overload  ...
```
```suggestion
Without changing the default behavior of the system, the load will not be distributed because it needs to be served by a single range. This behavior is not just temporary; certain users may consistently experience a high volume of activity compared to the average user. This can result in a system with multiple hotspots, each of which can potentially overload the system at any moment.
```

I'm not sure that defining the users is necessary.

florence-crl replied:
updated

kathancox commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+Because sequences avoid user expressions, optimizations can be made to improve their performance, but unfortunately the write volume on the sequence is still that of the sum total of all its accesses.
+
+[Sequence caching]({% link {{ page.version.version }}/create-sequence.md %}#cache-sequence-values-in-memory), which allows clients to cache sequence values as to reduce the burden on the target range, serves as a good mitigation for hot sequences. Alternatively, the `unique_rowid()` function generates sequential values which have strong guarantees against collision, with the drawback that its val ...
```
```suggestion
[Sequence caching]({% link {{ page.version.version }}/create-sequence.md %}#cache-sequence-values-in-memory), which allows clients to cache sequence values to reduce the burden on the target range, serves as a good mitigation for hot sequences. Alternatively, the `unique_rowid()` function generates sequential values which have strong guarantees against collision, with the drawback that its values are not a series.
```

florence-crl replied:
updated

kathancox commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+<img src="{{ 'images/v25.1/hotspots-figure-9.png' | relative_url }}" alt="Table hotspot example" style="border:1px solid #eee;max-width:100%" />
+
+Reads in the `posts` table may be evenly distributed, but joining the `countries` table becomes a bottleneck, since it exists in so few ranges. Splitting the `countries` table ranges can relieve pressure, but only to a theoretical limit as the indivisible points, the rows themselves, experience high throughput. [Global tables]({% link {{ page.version.version }}/global-tables.md %}) and [follower  ...
```
```suggestion
Reads in the `posts` table may be evenly distributed, but joining the `countries` table becomes a bottleneck, since it exists in so few ranges. Splitting the `countries` table ranges can relieve pressure, but only to a limit as the indivisible rows experience high throughput. [Global tables]({% link {{ page.version.version }}/global-tables.md %}) and [follower reads]({% link {{ page.version.version }}/follower-reads.md %}) can help scaling in this case, especially when write throughput is low.
```

florence-crl replied:
updated

kathancox commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+) STORED;
+~~~
+
+In the `ALTER` statement, the first condition `province <> 'alabama'` checks whether the province is not `alabama`. It matches every single row in the table that is not `alabama`, and will ironically place them in the `us-east-1` region.
```
My sense is that the "ironically" may be lost on some users?

florence-crl replied:
removed

kathancox commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+**Synonyms:** time-based hotspot
+
+_Temporal hotspots_ refer to increased database usage during particular windows of time. These take a variety of shapes, from event and holiday usage (such as Black Friday or the Superbowl), to synchronized job runs.
```
I was a bit split on the mentions of "Black Friday and the Superbowl". I think it's good to provide practical examples that users can understand, but my one worry was that they're both US-centric and I wonder whether they're timeless? Do you think just "from event and holiday usage" is enough without the parenthesis?

florence-crl replied:
removed

kathancox reviewed (commented):
Wow! I learned a lot from this. Thanks @florence-crl. I left a bunch of non-blocking comments, these are all suggestions/questions I have for clarification. Let me know if you want me to take another look.

florence-crl commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+Because [`CURRENT_TIMESTAMP()`]({% link {{ page.version.version }}/functions-and-operators.md %}#date-and-time-functions) is a steadily increasing value, this [`UPDATE`]({% link {{ page.version.version }}/update.md %}) will similarly burden the range at the tail of the index. While hotspots on an index tail tend to be the most common, bottlenecks on the head are not unheard of. For example, index ...
+
+In this page, the phrase _index hotspot_ will be reserved for a hot by write hotspot on an index, even though indexes can become hot due to read. This is because a hot by write index hotspot is the most common hotspot pattern that occurs now and in the future as workloads continue to be migrated from legacy single-node installations. Hot by read index hotspots are defined later on this page as [_ ...
```
updated

rmloveland commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+What makes lookback hotspots unique is that the cluster is not hot by write, but hot by read. Separately lookback hotspots also tend not to specify a key, which would evade systems using key requests to identify hotspots.
+
+#### Queuing Hotspot
```
oh nice, this is the queue-like workload section

suggest linking the `gc.ttlseconds` docs I mentioned earlier to this section

rmloveland commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+**Synonyms:** hot table
+
+_Hot tables_ are another variant of hot keys. Instead of a few select keys being burdened, the majority of the table is burdened by read (and at times write) access. This generally occurs on small reference tables that are used in queries with a join clause. This can lead to uneven key access.
```
suggest 'join clause' be a link to our docs on joins

rmloveland reviewed (approved):
LGTM! this doc is AWESOME!!!!! think this will be a game-changer

left some non-blocking comments for your consideration, feel free to take or leave

florence-crl reviewed (commented):
@kathancox thank you for the thorough first review! I have made your suggested updates. After I review and make updates suggested by @rmloveland, I will ask you to take another look

florence-crl commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+---
+title: Understand Hotspots
```
created follow-up issue: [DOC-12391](https://cockroachlabs.atlassian.net/browse/DOC-12391) Add link to hotspots doc from other related pages

florence-crl commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+All hotspot types described in this page will create hot nodes, as long as the cluster is not already operating at maximum capacity.
+
+The following image is a graph of [CPU Percent]({% link {{ page.version.version }}/ui-hardware-dashboard.md %}#cpu-percent) utilization by node. Most of the nodes hover around 25%, while one hot node is around 95%. The fact that the hot node changes indicates that the hotspot is moving as the [ranges]({% link {{ page.version.version }}/architecture/overview.md %}#range) containing writes fill up  ...
```
updated in previous commit

florence-crl commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+Identifying a hot node, or a node hotspot, is often part of troubleshooting hotspots in distributed databases. While this may not be the first place you notice issues, it is often the first place you can confidently identify a hotspot. Multiple nodes may become hot, but they will be clearly identifiable as their utilization metrics will be outliers compared to the rest of the cluster.
+
+All hotspot types described in this page will create hot nodes, as long as the cluster is not already operating at maximum capacity.
```
updated

florence-crl commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+All hotspot types described in this page will create hot nodes, as long as the cluster is not already operating at maximum capacity.
+
+The following image is a graph of [CPU Percent]({% link {{ page.version.version }}/ui-hardware-dashboard.md %}#cpu-percent) utilization by node. Most of the nodes hover around 25%, while one hot node is around 95%. The fact that the hot node changes indicates that the hotspot is moving as the [ranges]({% link {{ page.version.version }}/architecture/overview.md %}#range) containing writes fill up  ...
```
updated

florence-crl commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,313 @@
+
+A _write hotspot_ is a hotspot caused by write throughput. Write hotspots increase the likelihood of [contention]({% link {{ page.version.version }}/performance-best-practices-overview.md %}#transaction-contention) within the hot node or range, therefore leading to potential performance issues.
+
+Write hotspots also have the unique effect of affecting more than a single node. Since [consensus]({% link {{ page.version.version }}/architecture/overview.md %}#consensus) and [replication]({% link {{ page.version.version }}/architecture/overview.md %}#replication) must take place, write hotspots often affect three nodes (the default [replication factor]({% link {{ page.version.version }}/config ...
```
updated

florence-crl commented:
Updated with [reviewer_gpt_suggestions.md](https://github.com/user-attachments/files/18996001/reviewer_gpt_suggestions.md)

kathancox reviewed (approved):
LGTM! Awesome, love the changes you made @florence-crl , it's a great read. A couple of suggestions that are take or leave!

florence-crl commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,330 @@
+
+**Synonyms:** node hotspot
+
+A _hot node_ is the most common indicator of a hotspot, as the [DB Console Metrics page]({% link {{ page.version.version }}/ui-overview.md %}#metrics) shows resource utilization per node by default.
+
```
good catch! In my testing, the Docs Reviewer GPT recommended this as well.

florence-crl replied:
updated

kathancox replied:
Good to know I can catch what old Docs Reviewer catches 😄

florence-crl commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,330 @@
+
+**Synonyms:** index maximum present key
+
+An _index tail_ is the largest existing key of an index, lexigraphically. Note that this is different from the index (such as `/Table/1/1`) or index spans (such as `[/Table/1/1, /Table/1/2)`), because these are not valid row keys. An index tail instead refers to the last real key that appears in the index. For example, the head and tail of the index `/Table/1/1` are: `[/Table/1/1/aardvark, /Table ...
```
updated

florence-crl commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,330 @@
+Now the writes are distributed into the tails of the shards, rather than the tail of the whole index. This benefits write performance but makes reads more challenging. If you need to read a subset of the data, you will have to scan each shard of the index.
+
+{{site.data.alerts.callout_info}}
+Paradoxical performance implications
```
updated

florence-crl commented on src/current/v25.1/understand-hotspots.md:
```diff
@@ -0,0 +1,330 @@
+{{site.data.alerts.callout_info}}
+Paradoxical performance implications
+
+Index hotspots paradoxically increase the efficiency of CockroachDB from the standpoint of the number of CPU cycles per write. This is primarily due to [write compactions]({% link {{ page.version.version }}/architecture/storage-layer.md %}#compaction). As index hotspots continually move in a single direction, the [SST (sorted string table)]({% link {{ page.version.version }}/architecture/storage- ...
```
updated

florence-crl reviewed (commented):
tftr @kathancox!
